core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnDelta.scala

```diff
@@ -77,37 +80,106 @@ final class ColumnDelta extends ColumnFormatValue with Delta {
-          // release own buffer too and delta should be unusable now
-          release()
         }
+      } finally {
+        oldColumnValue.release()
+        newValue.release()
+        // release own buffer too and delta should be unusable now
+        release()
       }
     }
   }
 
+  private def mergeStats(oldBuffer: ByteBuffer, newBuffer: ByteBuffer,
+      schema: StructType): ByteBuffer = {
+    val numColumnsInStats = ColumnStatsSchema.numStatsColumns(schema.length)
+    val oldStatsRow = Utils.toUnsafeRow(oldBuffer, numColumnsInStats)
+    val newStatsRow = Utils.toUnsafeRow(newBuffer, numColumnsInStats)
+    val oldCount = oldStatsRow.getInt(ColumnStatsSchema.COUNT_INDEX_IN_SCHEMA)
+    val newCount = newStatsRow.getInt(ColumnStatsSchema.COUNT_INDEX_IN_SCHEMA)
+
+    val values = new Array[Any](numColumnsInStats)
+    val statsSchema = new Array[StructField](numColumnsInStats)
+    val nullCountField = StructField("nullCount", IntegerType)
+    values(ColumnStatsSchema.COUNT_INDEX_IN_SCHEMA) = oldCount + newCount
+    statsSchema(ColumnStatsSchema.COUNT_INDEX_IN_SCHEMA) =
+        StructField("count", IntegerType, nullable = false)
+    // non-generated code for evaluation since this is only for one row
+    // (besides binding to two separate rows will need custom code)
+    for (i <- schema.indices) {
+      val statsIndex = i * ColumnStatsSchema.NUM_STATS_PER_COLUMN + 1
+      val dataType = schema(i).dataType
+      val lowerExpr = BoundReference(statsIndex, dataType, nullable = true)
+      val upperExpr = BoundReference(statsIndex + 1, dataType, nullable = true)
+      val nullCountExpr = BoundReference(statsIndex + 2, IntegerType, nullable = true)
+      val ordering = TypeUtils.getInterpretedOrdering(dataType)
+
+      val oldLower = lowerExpr.eval(oldStatsRow)
+      val newLower = lowerExpr.eval(newStatsRow)
+      val oldUpper = upperExpr.eval(oldStatsRow)
+      val newUpper = upperExpr.eval(newStatsRow)
+      val oldNullCount = nullCountExpr.eval(oldStatsRow)
+      val newNullCount = nullCountExpr.eval(newStatsRow)
+
+      // Unlike normal < or > semantics, comparison against null on either
+      // side should return the non-null value or null if both are null.
+      // This is like Spark's Greatest/Least semantics. Likewise nullCount
+      // should return null only if both are null else skip the null one, if any,
+      // like the "sum" aggregate semantics (and unlike the SQL add semantics that
+      // returns null if either of the sides is null). The "AddStats" extension
+      // to Add operator is to encapsulate that behaviour.
+      val lower =
+        if (newLower == null) oldLower
+        else if (oldLower == null) newLower
+        else if (ordering.lt(oldLower, newLower)) oldLower else newLower
+      val upper =
+        if (newUpper == null) oldUpper
+        else if (oldUpper == null) newUpper
+        else if (ordering.lt(oldLower, newLower)) newLower else oldLower
```

Comment:
hi @sumwale, I think this comparison should be of upper values. Also if we are comparing, here the upper of both should preside. But I did not get why this comparison is needed?

Reply:
Yes, should be upper comparison -- thanks for spotting. Was having failure in a test due to the same.